### PR TITLE
[ADT] Make Zippy more iterator-like for lifetime safety

### PR DESCRIPTION
**Pull request by dwblaikie**
@geoffromer identifier/encountered a lifetime issue when using concat+zip, zip would return by value, concat would take references to that value and use them in its result after they had expired.

This is a common problem with range adapters and the lifetime of values.

But it's also non-conforming with the C++ iterator requirements, I think
- partly because op-> should be supported (which I haven't done here) and that basically has to return by pointer.

So the best thing is to stash a value in the iterator and return a pointer/reference to that.

(some context that may or may not be relevant to this part of the code may be in https://github.com/llvm/llvm-project/commit/981ce8fa15afa11d083033240edb1daff29081c7 )

### CODE REVIEW

File: llvm/include/llvm/ADT/STLExtras.h [hunk 1]
```diff
@@ -704,10 +704,12 @@ struct zip_common : public zip_traits<ZipType, ReferenceTupleType, Iters...> {
...
 
   std::tuple<Iters...> iterators;
+  mutable std::optional<value_type> value;
 
 protected:
-  template <size_t... Ns> value_type deref(std::index_sequence<Ns...>) const {
-    return value_type(*std::get<Ns>(iterators)...);
+  template <size_t... Ns> const value_type &deref(std::index_sequence<Ns...>) const {
+    value.emplace(*std::get<Ns>(iterators)...);
+    return *value;
```
Comment (kuhar) [inline llvm/include/llvm/ADT/STLExtras.h:712]: What's the difference between emplace and assignment here?

File: llvm/unittests/ADT/IteratorTest.cpp [hunk 2]
```diff
@@ -482,39 +482,26 @@ TEST(ZipIteratorTest, ZipEqualConstCorrectness) {
...
-  const std::vector<bool> c_nemesis = nemesis;
-
-  for (auto &&[a, b, c, d] : zip_equal(first, c_first, nemesis, c_nemesis)) {
+  for (auto &&[a, b] : zip_equal(first, c_first)) {
     a = 2;
-    c = true;
     static_assert(!IsConstRef<decltype(a)>);
     static_assert(IsConstRef<decltype(b)>);
-    static_assert(!IsBoolConstRef<decltype(c)>);
-    static_assert(IsBoolConstRef<decltype(d)>);
```
Comment (kuhar) [inline llvm/unittests/ADT/IteratorTest.cpp:494]: Why is this removed?

**Review by kuhar — COMMENTED** [review]
What's the impact on compilation times? If I'm reading this right, this will increase the size of Zippy to the size of the value type? For example, if I zip two vectors of `std::string`, the `value_type` will be something like `tuple<string&, string&>` ==> 16B? 

I think concat works around it without an extra member? https://github.com/llvm/llvm-project/blob/df0551298868b164197a4e54e9444120dc96ff53/llvm/include/llvm/ADT/STLExtras.h#L1068-L1089

> identifier/encountered a lifetime issue when using concat+zip, zip would return by value, concat would take references to that value and use them in its result after they had expired.

Could you share a repro? Last time I checked, `concat` also forwarded ranges to its storage, so I'm not exactly sure what the issue is.

**Comment by kuhar** [comment]
OK, so the issue is that the lifetime of objects created by dereferencing zip iterators can be shorter than the iterators themselves? Did I get it right?

> https://godbolt.org/z/aK8drvGvY

Thanks for the repro. Based on the error message, would it be something we could work around with some `if constexpr (is_rvalue_ref_v<...>)` in concat range instead?

I'm asking because IIRC returning values from the deref operator will become valid in C++20, so I wonder if we will have to fix it this way anyway later on.

**Comment by dwblaikie** [comment]
> OK, so the issue is that the lifetime of objects created by dereferencing zip iterators can be shorter than the iterators themselves? Did I get it right?

Yep, pretty much.

> Thanks for the repro. Based on the error message, would it be something we could work around with some `if constexpr (is_rvalue_ref_v<...>)` in concat range instead?

Possibly

> I'm asking because IIRC returning values from the deref operator will become valid in C++20, so I wonder if we will have to fix it this way anyway later on.

I had some vague sense of that too, but didn't find (admittedly didn't look terribly hard) details on this - do you happen to have pointers to the specifics here?

**Comment by kuhar** [comment]
> I had some vague sense of that too, but didn't find (admittedly didn't look terribly hard) details on this - do you happen to have pointers to the specifics here?

IANALL but I found these:
* https://en.cppreference.com/w/cpp/iterator/forward_iterator#:~:text=Unlike%20the%20LegacyForwardIterator%20requirements%2C%20the%20forward_iterator%20concept%20does%20not%20require%20dereference%20to%20return%20a%20reference.
* https://stackoverflow.com/a/51094221
* https://en.cppreference.com/w/cpp/named_req/Iterator and https://en.cppreference.com/w/cpp/iterator/iter_t (note that there's no arrow operator check or anything that attempts to take the address of a dereferenced iterator)

**Comment by geoffromer** [comment]
> > I'm asking because IIRC returning values from the deref operator will become valid in C++20, so I wonder if we will have to fix it this way anyway later on.
> 
> I had some vague sense of that too, but didn't find (admittedly didn't look terribly hard) details on this - do you happen to have pointers to the specifics here?

It's more of an addition than a change: the new iterator concepts like `std::forward_iterator` don't require `operator*` to return an lvalue, but the pre-C++20 "named requirements" like _ForwardIterator_ haven't changed, except that they're now called e.g. _LegacyForwardIterator_. See [P0022](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0022r2.html) for background. Note that it was written in 2016, so the details may have changed, but I believe it reflects the overall direction that was eventually adopted.

Ultimately, this isn't a question of whether the iterator is "valid", it's a question of what usages the iterator supports. Right now, the `zip` iterator doesn't support usages that require `operator*` to return an lvalue, which means it doesn't support being nested inside `concat`. Nothing in C++20 will cause those usages to start working; if you want them to work, you'll need to change the iterator to make them work.

**Comment by kuhar** [comment]
> Ultimately, this isn't a question of whether the iterator is "valid", it's a question of what usages the iterator supports. Right now, the `zip` iterator doesn't support usages that require `operator*` to return an lvalue, which means it doesn't support being nested inside `concat`. Nothing in C++20 will cause those usages to start working; if you want them to work, you'll need to change the iterator to make them work.

My point is that there are two things we could change: zippy or concat. We could teach concat to support iterators that don't satisfy this requirement and cover more range types in the same sweep as fixing issues with concat of zip.

**Comment by dwblaikie** [comment]
Yeah, fair - I'll look into what it'd take to make concat compatible with the `std::forward_iterator` concept in this regard.
